api/src/main/java/org/apache/iceberg/metrics/ScanReport.java

```diff
@@ -0,0 +1,251 @@
+    this.totalFileSizeBytes = totalFileSizeBytes;
+  }
+
+  public String tableName() {
+    return tableName;
+  }
+
+  public long snapshotId() {
+    return snapshotId;
+  }
+
+  public Expression filter() {
+    return filter;
+  }
+
+  public Schema projection() {
+    return projection;
+  }
+
+  public int matchingDataFiles() {
+    return matchingDataFiles;
+  }
+
+  public int matchingDataManifests() {
+    return matchingDataManifests;
+  }
+
+  public int totalDataManifestsRead() {
+    return totalDataManifestsRead;
+  }
+
+  public int addedDataFiles() {
+    return addedDataFiles;
+  }
+
+  public int deletedDataFiles() {
+    return deletedDataFiles;
+  }
+
+  public Duration totalPlanningDuration() {
+    return totalPlanningDuration;
+  }
+
+  public long totalFileSizeBytes() {
+    return totalFileSizeBytes;
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  @Override
+  public String toString() {
+    return MoreObjects.toStringHelper(this)
+        .add("tableName", tableName)
+        .add("snapshotId", snapshotId)
+        .add("filter", filter)
+        .add("projection", projection)
+        .add("matchingDataFiles", matchingDataFiles)
+        .add("matchingDataManifests", matchingDataManifests)
+        .add("totalDataManifestsRead", totalDataManifestsRead)
+        .add("addedDataFiles", addedDataFiles)
+        .add("deletedDataFiles", deletedDataFiles)
+        .add("totalScanDuration", totalPlanningDuration)
+        .add("totalFileSizeBytes", totalFileSizeBytes)
+        .toString();
+  }
+
+  @SuppressWarnings("HiddenField")
+  public static class Builder {
+    private String tableName;
+    private long snapshotId = -1L;
+    private Expression filter;
+    private Schema projection;
+    private ScanMetrics scanMetrics;
+
+    private Builder() {
+    }
+
+    public Builder withTableName(String tableName) {
+      this.tableName = tableName;
+      return this;
+    }
+
+    public Builder withSnapshotId(long snapshotId) {
+      this.snapshotId = snapshotId;
+      return this;
+    }
+
+    public Builder withFilter(Expression filter) {
+      this.filter = filter;
+      return this;
+    }
+
+    public Builder withProjection(Schema projection) {
+      this.projection = projection;
+      return this;
+    }
+
+    public Builder fromScanMetrics(ScanMetrics scanMetrics) {
+      this.scanMetrics = scanMetrics;
+      return this;
+    }
+
+    public ScanReport build() {
+      Preconditions.checkArgument(null != tableName, "TableName must be non-null");
+      Preconditions.checkArgument(null != filter, "Expression filter must be non-null");
+      Preconditions.checkArgument(null != projection, "Schema projection must be non-null");
+      Preconditions.checkArgument(null != scanMetrics, "ScanMetrics must be non-null");
+      return new ScanReport(tableName, snapshotId, filter, projection,
+          scanMetrics.matchingDataFiles().count().orElse(-1),
```

Comment:
Following up on this comment, I

Reply:
one of the reasons I added this distinction between `ScanMetrics` and the final `ScanReport` is so that we wouldn't need to serialize Counters/Timers. Additionally, it seemed cleaner to me to have metrics that are being updated wrapped in one class, where we would derive a final `ScanReport` from that is immutable (well kind-of). 
The `ScanReport` is just a  Data holder that is serializable, but it doesn't carry the baggage of whatever a particular `Counter`/`Timer` implementation would have.

I understand the reasoning behind making `Counter#count()` return an `Optional` but I was actually wondering whether it wouldn't be better to make this return an actual value. It seems overly restrictive to have this be an `Optional` for a special case, such as Hadoop. Also you generally don't see such a restriction in the Counter API of Micrometer. Maybe in such a case the `Counters` that are being produced by `HadoopMetricsContext` would just return -1 for `count()` if they can't fetch whatever they are counting from `FileSystem.Statistics`.
@danielcweeks thoughts on this?

Reply:
regarding `Counter#count()` being `Optional` I talked to @danielcweeks and we eventually concluded that it would probably be best to actually report the count and throw an exception in the Hadoop case. I opened https://github.com/apache/iceberg/pull/5328 for that

Reply:
I don't think that we need to serialize timers or counters if we don't have separate `ScanReport` and `ScanMetrics` classes. Scan planning should happen only on the coordinator, driver, or planner node. The resulting tasks are serialized to workers, but the scan itself should not be (and `TableScanContext` is not `Serializable`).

Reply:
I'm rather thinking of the case where someone would want to send the `ScanReport` over the wire. For example, with the REST catalog we might want to send the `ScanReport` to an endpoint via a `POST`. Having actual `Counters` / `Timers` in the `ScanReport` makes things difficult. Does that make sense?